api/src/main/java/io/minio/MinioClient.java

```diff
@@ -3895,6 +3902,201 @@ public ObjectLockConfiguration getDefaultRetention(String bucketName)
   }
 
 
+  /**
+   * Applies object retention lock onto an object.
+   *
+   * </p><b>Example:</b><br>
+   * <pre>{@code minioClient.setObjectRetention("my-bucketname", "my-object", config );
+   * System.out.println("Set object retention on my-object successfully."); }</pre>
+   *
+   * @param bucketName Bucket name.
+   * @param objectName Object name.
+   * @param versionId  Object versio id.
+   * @param bypassGovernanceRetention  By pass governance retention.
+   * @param config     Object lock configuration.
+   *
+   * @throws InvalidBucketNameException  upon invalid bucket name is given
+   * @throws NoSuchAlgorithmException
+   *           upon requested algorithm was not found during signature calculation
+   * @throws InsufficientDataException  upon getting EOFException while reading given
+   *           InputStream even before reading given length
+   * @throws IOException                 upon connection error
+   * @throws InvalidKeyException
+   *           upon an invalid access key or secret key
+   * @throws NoResponseException         upon no response from server
+   * @throws XmlPullParserException      upon parsing response xml
+   * @throws ErrorResponseException      upon unsuccessful execution
+   * @throws InternalException           upon internal library error
+   * @throws InvalidResponseException    upon a non-xml response from server
+   */
+  public void setObjectRetention(String bucketName, String objectName, String versionId, 
+          boolean bypassGovernanceRetention, ObjectRetentionConfiguration config)
+    throws InvalidBucketNameException, NoSuchAlgorithmException, InsufficientDataException, IOException,
+           InvalidKeyException, NoResponseException, XmlPullParserException, ErrorResponseException,
+           InternalException, InvalidResponseException {
+
+    Map<String, String> queryParamMap = new HashMap<>();
+    queryParamMap.put("retention", "");
+
+    if (!versionId.isEmpty()) {
+      queryParamMap.put("versionId", versionId);
+    }
+
+    Map<String, String> headerMap = new HashMap<>();
+    if (bypassGovernanceRetention) {
+      headerMap.put("x-amz-bypass-governance-retention", "True");
+    }
+
+    HttpResponse response = executePut(bucketName, objectName, headerMap, queryParamMap, config, 0);
```

Comment:
`config` must be non-null and `Mode` and `RetainUntilDate` are mandatory configuration. You would need to add this validation properly.